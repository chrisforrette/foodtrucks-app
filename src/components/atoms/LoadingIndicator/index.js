import React from 'react'
import PropTypes from 'prop-types'
import {
  default as styled,
  keyframes
} from 'styled-components'

const spin = keyframes`
  0% {
    transform: rotate(0deg);
  }

  100% {
    transform: rotate(360deg);
  }
`

const Wrapper = styled.div`
  animation: ${spin} 1s linear infinite;
  height: ${({ size }) => size}px;
  line-height: 0;
  opacity: ${({ isLoading }) => isLoading ? 1 : 0};
  ${({ fade }) => fade && 'transition: opacity 250ms ease;'}
  width: ${({ size }) => size}px;
`

const createSvgCircle = style => (
  <circle cx="16" cy="16" fill="none" r="14" strokeWidth="4" style={style} />
)

const LoadingIndicator = props => {
  const { color } = props
  return <Wrapper {...props}>
    <svg height='100%' viewBox='0 0 32 32' width='100%'>
      {createSvgCircle({
        stroke: color,
        opacity: 0.25
      })}
      {createSvgCircle({
        stroke: color,
        strokeDasharray: 80,
        strokeDashoffset: 55
      })}
    </svg>
  </Wrapper>
}

LoadingIndicator.defaultProps = {
  color: '#fff',
  fade: false,
  isLoading: false,
  size: 16
}

LoadingIndicator.propTypes = {
  color: PropTypes.string,
  fade: PropTypes.bool,
  isLoading: PropTypes.bool,
  size: PropTypes.number
}

export default LoadingIndicator
