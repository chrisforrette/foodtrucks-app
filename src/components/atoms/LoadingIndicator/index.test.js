import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

import LoadingIndicator from './index'

it('should render a loading indicator', () => {
  const rendered = shallow(<LoadingIndicator />)
  expect(toJson(rendered)).toMatchSnapshot()
})

it('should include a fade transition when fade is true', () => {
  const rendered = shallow(<LoadingIndicator fade />)
  expect(toJson(rendered)).toMatchSnapshot()
})

it('should be visible when loading is true', () => {
  const rendered = shallow(<LoadingIndicator loading />)
  expect(toJson(rendered)).toMatchSnapshot()
})

it('should accept a size', () => {
  const rendered = shallow(<LoadingIndicator size={32} />)
  expect(toJson(rendered)).toMatchSnapshot()
})

it('should accept a color', () => {
  const rendered = shallow(<LoadingIndicator color='#f60' />)
  expect(toJson(rendered)).toMatchSnapshot()
})
