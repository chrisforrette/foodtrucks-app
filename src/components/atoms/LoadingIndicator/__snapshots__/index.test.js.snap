// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should accept a color 1`] = `
<styled.div
  color="#f60"
  fade={false}
  isLoading={false}
  size={16}
>
  <svg
    height="100%"
    viewBox="0 0 32 32"
    width="100%"
  >
    <circle
      cx="16"
      cy="16"
      fill="none"
      r="14"
      strokeWidth="4"
      style={
        Object {
          "opacity": 0.25,
          "stroke": "#f60",
        }
      }
    />
    <circle
      cx="16"
      cy="16"
      fill="none"
      r="14"
      strokeWidth="4"
      style={
        Object {
          "stroke": "#f60",
          "strokeDasharray": 80,
          "strokeDashoffset": 55,
        }
      }
    />
  </svg>
</styled.div>
`;

exports[`should accept a size 1`] = `
<styled.div
  color="#fff"
  fade={false}
  isLoading={false}
  size={32}
>
  <svg
    height="100%"
    viewBox="0 0 32 32"
    width="100%"
  >
    <circle
      cx="16"
      cy="16"
      fill="none"
      r="14"
      strokeWidth="4"
      style={
        Object {
          "opacity": 0.25,
          "stroke": "#fff",
        }
      }
    />
    <circle
      cx="16"
      cy="16"
      fill="none"
      r="14"
      strokeWidth="4"
      style={
        Object {
          "stroke": "#fff",
          "strokeDasharray": 80,
          "strokeDashoffset": 55,
        }
      }
    />
  </svg>
</styled.div>
`;

exports[`should be visible when loading is true 1`] = `
<styled.div
  color="#fff"
  fade={false}
  isLoading={false}
  loading={true}
  size={16}
>
  <svg
    height="100%"
    viewBox="0 0 32 32"
    width="100%"
  >
    <circle
      cx="16"
      cy="16"
      fill="none"
      r="14"
      strokeWidth="4"
      style={
        Object {
          "opacity": 0.25,
          "stroke": "#fff",
        }
      }
    />
    <circle
      cx="16"
      cy="16"
      fill="none"
      r="14"
      strokeWidth="4"
      style={
        Object {
          "stroke": "#fff",
          "strokeDasharray": 80,
          "strokeDashoffset": 55,
        }
      }
    />
  </svg>
</styled.div>
`;

exports[`should include a fade transition when fade is true 1`] = `
<styled.div
  color="#fff"
  fade={true}
  isLoading={false}
  size={16}
>
  <svg
    height="100%"
    viewBox="0 0 32 32"
    width="100%"
  >
    <circle
      cx="16"
      cy="16"
      fill="none"
      r="14"
      strokeWidth="4"
      style={
        Object {
          "opacity": 0.25,
          "stroke": "#fff",
        }
      }
    />
    <circle
      cx="16"
      cy="16"
      fill="none"
      r="14"
      strokeWidth="4"
      style={
        Object {
          "stroke": "#fff",
          "strokeDasharray": 80,
          "strokeDashoffset": 55,
        }
      }
    />
  </svg>
</styled.div>
`;

exports[`should render a loading indicator 1`] = `
<styled.div
  color="#fff"
  fade={false}
  isLoading={false}
  size={16}
>
  <svg
    height="100%"
    viewBox="0 0 32 32"
    width="100%"
  >
    <circle
      cx="16"
      cy="16"
      fill="none"
      r="14"
      strokeWidth="4"
      style={
        Object {
          "opacity": 0.25,
          "stroke": "#fff",
        }
      }
    />
    <circle
      cx="16"
      cy="16"
      fill="none"
      r="14"
      strokeWidth="4"
      style={
        Object {
          "stroke": "#fff",
          "strokeDasharray": 80,
          "strokeDashoffset": 55,
        }
      }
    />
  </svg>
</styled.div>
`;
